common/types/malfeasance.go

```diff
@@ -0,0 +1,159 @@
+package types
+
+import (
+	"errors"
+
+	"github.com/spacemeshos/go-scale"
+
+	"github.com/spacemeshos/go-spacemesh/codec"
+	"github.com/spacemeshos/go-spacemesh/log"
+)
+
+//go:generate scalegen -types MalfeasanceProof,MalfeasanceGossip,AtxProof,BallotProof,HareProof,AtxProofMsg,BallotProofMsg,HareProofMsg,HareMetadata
+
+const (
+	MultipleATXs uint8 = iota + 1
+	MultipleBallots
+	HareEquivocation
+)
+
+type MalfeasanceProof struct {
+	// for network upgrade
+	Layer     LayerID
+	ProofData TypedProof
+}
+
+type TypedProof struct {
+	// MultipleATXs | MultipleBallots | HareEquivocation
+	Type uint8
+	// AtxProof | BallotProof | HareProof
+	Proof scale.Type
+}
```

Comment:
Just a suggestion for a different naming and using a custom type for the enum:

```suggestion
type ProofType byte

const (
	MultipleATXs ProofType = iota + 1
	MultipleBallots
	HareEquivocation
)

type MalfeasanceProof struct {
	// for network upgrade
	Layer LayerID
	Proof Proof
}

type Proof struct {
	// MultipleATXs | MultipleBallots | HareEquivocation
	Type ProofType
	// AtxProof | BallotProof | HareProof
	Data scale.Type
}
```

Reply:
i did have a custom type, but in the code i have to constantly cast it back and forth.

Reply:
renamed as suggested, with the exception of the custom type